lib/internal/modules/helpers.js

```diff
@@ -300,31 +301,28 @@ function normalizeReferrerURL(referrer) {
 }
 
 /**
- * For error messages only, check if ESM syntax is in use.
- * @param {string} code
+ * Check if the error is a syntax error due to ESM syntax in CommonJS.
+ * - `import` statements return an error with a message `Cannot use import statement outside a module`.
+ * - `export` statements return an error with a message `Unexpected token 'export'`.
+ * - `import.meta` returns an error with a message `Cannot use 'import.meta' outside a module`.
+ * Top-level `await` currently returns the same error message as when `await` is used in a sync function,
+ * so we don't use it as a disambiguation.
+ * Dynamic `import()` is permitted in CommonJS, so we don't use it as a disambiguation.
+ * @param {Error} err
  */
-function hasEsmSyntax(code) {
-  debug('Checking for ESM syntax');
-  const parser = require('internal/deps/acorn/acorn/dist/acorn').Parser;
-  let root;
-  try {
-    root = parser.parse(code, { sourceType: 'module', ecmaVersion: 'latest' });
-  } catch {
-    return false;
-  }
-
-  return ArrayPrototypeSome(root.body, (stmt) =>
-    stmt.type === 'ExportDefaultDeclaration' ||
-    stmt.type === 'ExportNamedDeclaration' ||
-    stmt.type === 'ImportDeclaration' ||
-    stmt.type === 'ExportAllDeclaration');
+function isModuleSyntaxError(err) {
+  return err != null && ObjectGetPrototypeOf(err) === SyntaxErrorPrototype && (
```

Comment:
should this use `utils.isError` so that it can't be tricked by `{ __proto__: SyntaxError.prototype }`?

Reply:
Note that it was already there before this PR, modifying it should be a separate discussion imo.